### PR TITLE
Issue #86 Add a callback to filter sections from course format

### PR DESCRIPTION
**Pull request by TomoTsuyuki**
Implement callback function `massaction_restricted_sections` to filter sections for preventing move/copy.

Course format plugin can implement **_pluginname_**_massaction_restricted_sections in their `lib.php` and filter the sections from the course format rule.

The parameter for the **_pluginname_**_massaction_restricted_sections is courseid, and returns array of sections.

### CODE REVIEW

**Comment by TomoTsuyuki** [comment]
This is for issue #86

**Comment by TomoTsuyuki** [comment]
Rebased with the latest `master` branch.

File: classes/actions.php [hunk 1]
```diff
@@ -507,6 +527,11 @@ public static function perform_moveto(array $modules, int $target): void {
                 throw new moodle_exception('sectionnotexist', 'block_massaction');
             }
 
+            // Not moving if the section is restricted.
+            if (in_array($cm->sectionnum, $sectionsrestricted)) {
+                continue;
```
Comment (Syxton) [inline classes/actions.php:532]: Should we throw an exception of sectionrestricted here?

File: classes/actions.php [hunk 2]
```diff
@@ -262,8 +270,15 @@ public static function duplicate_to_course(array $modules, int $targetcourseid,
...
         // mod1, mod2, mod3, mod4, mod5, mod2(dup), mod4(dup).
         $duplicatedmods = [];
+        $sourceformat = course_get_format($sourcecourseid);
+        $sourcesectionsrestricted = massactionutils::get_restricted_sections($sourcecourseid, $sourceformat->get_format());
         foreach ($idsincourseorder as $cmid) {
-            $duplicatedmod = massactionutils::duplicate_cm_to_course($targetmodinfo->get_course(), $sourcemodinfo->get_cm($cmid));
+            $sourcecm = $sourcemodinfo->get_cm($cmid);
+            // Not duplicated if the section is restricted.
+            if (in_array($sourcecm->sectionnum, $sourcesectionsrestricted)) {
+                continue;
```
Comment (Syxton) [inline classes/actions.php:279]: Should we throw an exception of sectionrestricted here?

File: classes/actions.php [hunk 3]
```diff
@@ -147,7 +147,15 @@ public static function duplicate(array $modules, $sectionnumber = false): void {
...
         // Let order of mods in a section be mod1, mod2, mod3, mod4, mod5. If we duplicate mod2, mod4, the order afterwards will be
         // mod1, mod2, mod3, mod4, mod5, mod2(dup), mod4(dup).
+        $duplicatedmods = [];
+        $targetformat = course_get_format($courseid);
+        $sectionsrestricted = massactionutils::get_restricted_sections($courseid, $targetformat->get_format());
         foreach ($idsincourseorder as $cmid) {
+            $cm = $modinfo->get_cm($cmid);
+            // Not duplicated if the section is restricted.
+            if (in_array($cm->sectionnum, $sectionsrestricted)) {
+                continue;
```
Comment (Syxton) [inline classes/actions.php:157]: Should we throw an exception of sectionrestricted here?

**Review by Syxton — CHANGES_REQUESTED** [review]
I like these changes, but it seems we might want to throw an exception if we somehow get to a restricted section instead of skipping past it.  What do you think?

**Pull request merged**